File: app/static/js/integrations.js
/* Integrations & privacy (task-36). Renders /api/integrations/sources
   verbatim: every status/date/chip is computed server-side (see that
   route's docstring for the exact "connected/stale/none" rules), so this
   file only builds row text/markup — no status math here. */
(function () {
  if (!window.HermesCharts) return;
  const { fetchJSON } = window.HermesCharts;

  const FLAG_CLASS = { connected: "in", stale: "bd", none: "bd", retired: "bd", not_running: "out", unknown: "bd" };

  function row(name, chipText, flagClass, sub) {
    const el = document.createElement("div");
    el.className = "src-row";
    const left = document.createElement("div");
    const nameEl = document.createElement("div");
    nameEl.className = "strong";
    nameEl.textContent = name;
    left.appendChild(nameEl);
    if (sub) {
      const subEl = document.createElement("div");
      subEl.className = "muted micro mt-1";
      subEl.textContent = sub;
      left.appendChild(subEl);
    }
    const chip = document.createElement("span");
    chip.className = "flag " + flagClass;
    chip.textContent = chipText;
    el.append(left, chip);
    return el;
  }

  // "Connected · last sync <date>" / "Stale · last sync <date>" / "No data
  // yet" — the shared chip-text shape for the two date-backed sources.
  function syncChip(status, date) {
    if (status === "connected") return `Connected · last sync ${date}`;
    if (status === "stale") return `Stale · last sync ${date}`;
    return "No data yet";
  }

  function telegramRow(t) {
    t = t || {};
    const chipText = { connected: "Connected", not_running: "Not running" }[t.status] || "Status unknown";
    let sub;
    if (t.status === "connected") sub = [t.model, t.provider].filter(Boolean).join(" · ") || null;
    else sub = t.error || null;
    return row("Telegram bot", chipText, FLAG_CLASS[t.status] || "bd", sub);
  }

  function hevyRow(h) {
    h = h || {};
    return row("Hevy (training)", syncChip(h.status, h.latest_date), FLAG_CLASS[h.status] || "bd");
  }

  function wearableRow(w) {
    w = w || {};
    const apple = w.apple || {};
    // Apple is retired by product contract and no longer feeds
    // w.status/latest_date — it's shown as its own honest, non-error state.
    const appleBit = apple.since
      ? `Apple: retired ${apple.since} · archive imports occasionally`
      : "Apple: retired";
    const sub = `${appleBit} · Fitbit: ${w.fitbit_latest || "no data"}` +
      ` (token: ${w.fitbit_auth_status || "unknown"})`;
    return row("Wearable", syncChip(w.status, w.latest_date), FLAG_CLASS[w.status] || "bd", sub);
  }

  function labsRow(l) {
    l = l || {};
    const sub = l.latest_date ? `latest result: ${l.latest_date}` : "no results logged yet";
    return row("Bloodwork lab", "Manual upload", "bd", sub);
  }

  async function boot() {
    const list = document.getElementById("src-list");
    let data;
    try {
      data = await fetchJSON("/api/integrations/sources");
    } catch (e) {
      const p = document.createElement("p");
      p.className = "muted note-txt";
      p.textContent = "Sources unavailable: " + e.message;
      list.appendChild(p);
      return;
    }
    list.append(
      telegramRow(data.telegram),
      hevyRow(data.hevy),
      wearableRow(data.wearable),
      labsRow(data.labs),
    );
  }
  boot();
})();

File: app/routes/integrations.py
"""Integrations & privacy settings page (task-36): a thin, read-only status
board over the sources the vault already integrates with. No new bridge
subcommands — this composes db_read SELECTs with the existing hermes-status
and fitbit-status bridge reads (both already ALLOWED and panel-reachable via
/api/models/*; see app/bridge.py).

Every chip below is a DETERMINISTIC derivation, documented here, never a
hardcoded "Connected" — stale or missing data says so (owner ground rule).
"""
from flask import Blueprint, jsonify

from app import bridge, canon, db_read

bp = Blueprint("integrations", __name__, url_prefix="/api/integrations")

# "Connected" freshness windows, in days. The only two rows here with a data
# DATE to compare against "now" (Telegram is a live gateway health check, not
# a date; Bloodwork is manual-upload by design, so its date is informational
# only — see below). Hevy: 2-3 sessions/week is normal, so a run without a
# hard rest week still reads Connected at 14 days. Wearable: Fitbit syncs
# roughly daily, so 3 days catches a real sync outage without flagging a
# single day the phone wasn't opened.
HEVY_FRESH_DAYS = 14
WEARABLE_FRESH_DAYS = 3

# Product decision: the Apple wearable pipeline is retired (device
# replaced; last apple daily_metrics row 2026-05-31). Apple's MAX(date) is
# frozen going forward, so it must never drag or prop up the live wearable
# status — occasional manual archive imports may still land (roughly
# monthly/quarterly), so apple_latest keeps reading the real DB date.
RETIRED_SOURCES = {"apple": "2026-05-31"}


def _guard():
    if not db_read.available():
        return jsonify(error="health database not available"), 503
    return None


def _max_date(sql):
    rows = db_read.query(sql)
    return rows[0]["m"] if rows and rows[0]["m"] else None


def _status_for(latest_date, fresh_days):
    """connected/stale/none against a fresh-days cutoff, computed on the
    canonical configured day (app.canon) — never SQLite's UTC 'now'."""
    if not latest_date:
        return "none"
    return "connected" if latest_date >= canon.days_ago_iso(fresh_days) else "stale"


def _telegram_status(agent):
    if "error" in agent:
        return "unknown"
    if agent.get("gateway_active") is True:
        return "connected"
    if agent.get("gateway_active") is False:
        return "not_running"
    return "unknown"


@bp.get("/sources")
def sources():
    """Composes the Connected-sources card (design image 31): Telegram bot /
    Hermes agent (live gateway check via hermes-status), Hevy training sync
    (last hevy_sets date), Wearable (Fitbit daily_metrics date + auth state —
    Apple is retired per RETIRED_SOURCES and reported separately so it can
    never drag or prop up the live status), Bloodwork lab (latest labs date —
    manual-upload by design, so it gets no freshness judgment, only the real
    date if any)."""
    if (resp := _guard()) is not None:
        return resp

    try:
        agent = bridge.run("hermes-status")
    except bridge.BridgeError as exc:
        agent = {"error": str(exc)}

    try:
        fitbit = bridge.run("fitbit-status")
    except bridge.BridgeError as exc:
        fitbit = {"error": str(exc)}

    hevy_latest = _max_date("SELECT MAX(date) AS m FROM hevy_sets")
    apple_latest = _max_date("SELECT MAX(date) AS m FROM daily_metrics WHERE source = 'apple'")
    fitbit_latest = _max_date("SELECT MAX(date) AS m FROM daily_metrics WHERE source = 'fitbit'")
    labs_latest = _max_date("SELECT MAX(date) AS m FROM labs")
    # Apple is retired (see RETIRED_SOURCES) — the combined wearable status
    # comes from fitbit alone now, even when apple rows exist and are newer.
    wearable_latest = fitbit_latest

    return jsonify(
        telegram={
            "status": _telegram_status(agent),
            "model": agent.get("model"),
            "provider": agent.get("provider"),
            "error": agent.get("error"),
        },
        hevy={
            "status": _status_for(hevy_latest, HEVY_FRESH_DAYS),
            "latest_date": hevy_latest,
        },
        wearable={
            "status": _status_for(wearable_latest, WEARABLE_FRESH_DAYS),
            "latest_date": wearable_latest,
            "apple_latest": apple_latest,
            "fitbit_latest": fitbit_latest,
            "fitbit_auth_status": fitbit.get("auth_status"),
            "apple": {
                "status": "retired",
                "latest_date": apple_latest,
                "since": RETIRED_SOURCES["apple"],
                "note": f"retired {RETIRED_SOURCES['apple']} — wearable replaced; "
                        "archive imports may still land occasionally",
            },
        },
        labs={
            "latest_date": labs_latest,
        },
    )
